thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/AbstractAnomalyDetectionFunction.java

```diff
@@ -0,0 +1,67 @@
+package com.linkedin.thirdeye.anomalydetection.function;
+
+import com.linkedin.thirdeye.anomalydetection.Utils;
+import com.linkedin.thirdeye.anomalydetection.model.data.DataModel;
+import com.linkedin.thirdeye.anomalydetection.model.data.NoopDataModel;
+import com.linkedin.thirdeye.anomalydetection.model.detection.DetectionModel;
+import com.linkedin.thirdeye.anomalydetection.model.detection.NoopDetectionModel;
+import com.linkedin.thirdeye.anomalydetection.model.prediction.NoopPredictionModel;
+import com.linkedin.thirdeye.anomalydetection.model.prediction.PredictionModel;
+import com.linkedin.thirdeye.anomalydetection.model.transform.TransformationFunction;
+import com.linkedin.thirdeye.datalayer.dto.AnomalyFunctionDTO;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+import org.joda.time.Interval;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractAnomalyDetectionFunction implements AnomalyDetectionFunction {
+  protected final Logger LOGGER = LoggerFactory.getLogger(this.getClass());
+
+  protected AnomalyFunctionDTO spec;
+  protected Properties properties;
+
+  protected DataModel dataModel = new NoopDataModel();
+  protected List<TransformationFunction> currentTimeSeriesTransformationChain = new ArrayList<>();
+  protected List<TransformationFunction> baselineTimeSeriesTransformationChain = new ArrayList<>();
```

Comment:
please take a look at "chain of responsibility" design pattern and how we can trigger "apply" method one after another.